File: Tetris/Golang/game.go
package main

import (
	rl "github.com/gen2brain/raylib-go/raylib"
	"time"
)

type Game struct {
	windowWidth         int32
	windowHeight        int32
	cellSize            int32
	targetFPS           int32
	title               string
	rows                int32
	columns             int32
	forcedMovementDelta time.Duration
	colors              []rl.Color
	grid                Grid
	movementTick        time.Time
	downMovement        time.Time
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) init() {
	g.windowHeight = g.rows * g.cellSize
	g.windowWidth = g.columns * g.cellSize
	rl.InitWindow(g.windowWidth, g.windowHeight, g.title)
	rl.SetTargetFPS(g.targetFPS)

	g.loadColors()
	g.createGrid()
	g.forcedMovementDelta = 1 * time.Second
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) autoMoveDown() {
	delta := time.Now().Sub(g.downMovement)

	if delta > g.forcedMovementDelta {
		g.grid.moveBlock("down")
		g.downMovement = time.Now()
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) checkNewBlock() {
	for _, cell := range g.grid.currentBlock[g.grid.rotation] {
		y := cell[1]
		if y+g.grid.yOffset == g.rows-1 {
			g.grid.selectBlock()
		}
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) checkCollisionX(direction int32) bool {
	positions := g.grid.blockCoordinates()

	for _, cell := range g.grid.currentBlock[g.grid.rotation] {
		if g.grid.checkSelfX(cell[0], cell[1], direction, positions) {
			continue
		}
		nextX := cell[0] + g.grid.xOffset + direction
		nextY := cell[1] + g.grid.yOffset

		if nextX < 0 || nextX > g.columns-1 {
			return false
		}

		if g.grid.cells[nextY][nextX] != 0 {
			return false
		}
	}
	return true
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) collisionY() {
	positions := g.grid.blockCoordinates()
	newBlock := false

	for _, cell := range g.grid.currentBlock[g.grid.rotation] {
		if g.grid.checkSelfY(cell[0], cell[1], positions) {
			continue
		}

		nextX := cell[0] + g.grid.xOffset
		nextY := cell[1] + g.grid.yOffset + 1

		if nextY > g.rows-1 {
			break
		}

		if g.grid.cells[nextY][nextX] != 0 {
			newBlock = true
		}
	}
	if newBlock {
		g.grid.selectBlock()
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) checkValidMoveLeft() bool {
	for _, cell := range g.grid.currentBlock[g.grid.rotation] {
		x := cell[0]
		if x+g.grid.xOffset-1 < 0 {
			return false
		}
		if !g.checkCollisionX(-1) {
			return false
		}
	}
	return true
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) checkValidMoveRight() bool {
	for _, cell := range g.grid.currentBlock[g.grid.rotation] {
		x := cell[0]
		if x+g.grid.xOffset+1 > g.columns-1 {
			return false
		}
		if !g.checkCollisionX(1) {
			return false
		}
	}
	return true
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) checkValidRotation() bool {
	nextRotation := g.grid.rotation + 1
	maxRotations := int32(3)
	if nextRotation > maxRotations {
		nextRotation = int32(0)
	}

	for _, cell := range g.grid.currentBlock[nextRotation] {
		x := cell[0]
		if x+g.grid.xOffset < 0 {
			return false
		}
		if x+g.grid.xOffset > g.columns-1 {
			return false
		}
	}
	return true
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) createGrid() {
	g.grid = Grid{
		cellSize: g.cellSize,
		rows:     g.rows,
		columns:  g.columns,
		colors:   g.colors,
	}
	g.grid.init()
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) increaseForcedMovementSpeed() {
	if g.grid.totalLinesCleared >= 1 {
		g.forcedMovementDelta /= 100
		g.forcedMovementDelta *= 90
		g.grid.totalLinesCleared = 0
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) loadColors() {
	g.colors = []rl.Color{
		rl.DarkGray,
		rl.Green,
		rl.Red,
		rl.Orange,
		rl.Yellow,
		rl.Purple,
		rl.SkyBlue,
		rl.Blue,
	}
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) run() {
	for !rl.WindowShouldClose() {
		g.collisionY()
		g.checkNewBlock()
		g.userInput()
		g.autoMoveDown()
		g.grid.updateGrid()
		g.increaseForcedMovementSpeed()

		rl.BeginDrawing()
		rl.ClearBackground(rl.Black)
		g.grid.drawGrid()
		rl.EndDrawing()
	}
	rl.CloseWindow()
}

// ---------------------------------------------------------------------------------------------------------------------

func (g *Game) userInput() {
	delta := time.Now().Sub(g.movementTick)

	if rl.IsKeyDown(rl.KeyLeft) && delta.Seconds() > 0.1 && g.checkValidMoveLeft() {
		g.grid.moveBlock("left")
		g.movementTick = time.Now()
	}
	if rl.IsKeyDown(rl.KeyRight) && delta.Seconds() > 0.1 && g.checkValidMoveRight() {
		g.grid.moveBlock("right")
		g.movementTick = time.Now()
	}
	if rl.IsKeyDown(rl.KeyDown) && delta.Seconds() > 0.1 {
		g.grid.moveBlock("down")
		g.movementTick = time.Now()
	}

	maxRotation := int32(3)
	if rl.IsKeyPressed(rl.KeyUp) && g.checkValidRotation() {
		g.grid.clearPrevious()
		g.grid.rotation++
		if g.grid.rotation > maxRotation {
			g.grid.rotation = 0
		}
	}
}

// ---------------------------------------------------------------------------------------------------------------------

File: Tetris/Python/game.py
from grid import *
from blocks import *
from ui import *
from sounds import *
from datetime import datetime, timedelta
import random
import pyray as pr


# ----------------------------------------------------------------------------------------------------------------------


class Game:
    def __init__(self) -> None:
        self.window_width = 500
        self.window_height = 620
        self.title = "Tetris"
        self.background_color = pr.Color(44, 44, 127, 255)
        self.targetFPS = 144
        self.setup_window()

        self.sound = Sounds()

        self.ui = UI()

        self.cell_size = 30
        self.grid = Grid(self.cell_size)

        self.blocks: list[Block] = self.fill_block_list()
        self.current_block: Block = self.get_block()
        self.next_block: Block = self.get_block()

        self.time_delta = 1.0
        self.auto_move_time: datetime = datetime.now()
        self.calculate_move_time()

        self.move_delta = 0.125
        self.next_move_delta: datetime = datetime.now()
        self.calculate_move_delta()

        self.game_over = False
        self.score = 0
        self.rows_cleared = 0

    # -------------------------------------------------------------------------

    def automatic_move_down(self) -> None:
        """Moves the block down after X amount of time has passed."""

        now: datetime = datetime.now()
        if now >= self.auto_move_time:
            self.move_block_down()
            self.calculate_move_time()

    # -------------------------------------------------------------------------

    def block_fits(self) -> bool:
        """Checks if the block fits in the grid after X movement."""

        cells: list[Position] = self.current_block.get_cell_positions()

        for cell in cells:
            if not self.grid.is_cell_empty(cell.row, cell.column):
                return False

        return True

    # -------------------------------------------------------------------------

    def calculate_move_delta(self) -> None:
        """Dictates time between user inputs."""

        self.next_move_delta: datetime = datetime.now() + timedelta(
            seconds=self.move_delta
        )

    # -------------------------------------------------------------------------

    def calculate_move_time(self) -> None:
        """Calculates the time in the future that the next automatic down
        movement should take place."""

        self.auto_move_time: datetime = datetime.now() + timedelta(
            seconds=self.time_delta
        )

    # -------------------------------------------------------------------------

    def draw(self) -> None:
        self.grid.draw_grid()
        self.current_block.draw(11, 11)
        self.ui.draw(self.game_over, self.score, self.next_block)

    # -------------------------------------------------------------------------

    def fill_block_list(self) -> list[Block]:
        """Returns a list of valid blocks to the caller."""
        return [
            IBlock(),
            JBlock(),
            LBlock(),
            OBlock(),
            SBlock(),
            TBlock(),
            ZBlock(),
        ]

    # -------------------------------------------------------------------------

    def get_block(self) -> Block:
        """Returns a random block to the caller."""

        choice = random.randint(0, len(self.blocks) - 1)
        block = self.blocks[choice]
        del self.blocks[choice]

        if len(self.blocks) == 0:
            self.blocks = self.fill_block_list()

        return block

    # -------------------------------------------------------------------------

    def handle_input(self) -> None:
        """Handles user input"""

        key_pressed: int = pr.get_key_pressed()
        if self.game_over and key_pressed:
            self.game_over = False
            self.reset()

        match key_pressed:
            case pr.KeyboardKey.KEY_UP:
                self.rotate_block()
            case pr.KeyboardKey.KEY_SPACE:
                self.snap_block_down()
            case _:
                pass

        now: datetime = datetime.now()

        if pr.is_key_down(pr.KeyboardKey.KEY_LEFT) and now > self.next_move_delta:
            self.move_block_left()
            self.calculate_move_delta()
        if pr.is_key_down(pr.KeyboardKey.KEY_RIGHT) and now > self.next_move_delta:
            self.move_block_right()
            self.calculate_move_delta()
        elif pr.is_key_down(pr.KeyboardKey.KEY_DOWN) and now > self.next_move_delta:
            self.move_block_down()
            self.calculate_move_delta()
            self.update_score(0, 1)

    # -------------------------------------------------------------------------

    def is_block_outside_grid(self) -> bool:
        """Checks if any of the cells, when moved would be outside the grid."""

        cells: list[Position] = self.current_block.get_cell_positions()

        for cell in cells:
            if self.grid.is_cell_outside_grid(cell.row, cell.column):
                return True

        return False

    # -------------------------------------------------------------------------

    def lock_block(self) -> None:
        """Locks the block in place by setting the color on the grid to the
        same as that block and selects a new block."""

        current_block: list[Position] = self.current_block.get_cell_positions()

        for cell in current_block:
            self.grid.grid[cell.row][cell.column] = self.current_block.id_code

        self.current_block = self.next_block
        if not self.block_fits():
            self.game_over = True

        self.next_block = self.get_block()

        rows_cleared: int = self.grid.clear_full_rows()

        if rows_cleared > 0:
            self.sound.play_clear_sound()
            self.update_score(rows_cleared, 0)
            self.rows_cleared += rows_cleared

        if self.rows_cleared >= 3:
            self.time_delta *= 0.9
            self.rows_cleared = 0

    # -------------------------------------------------------------------------

    def move_block_down(self) -> None:
        """Moves the block down 1 square."""

        if not self.game_over:
            self.current_block.move(1, 0)

            if self.is_block_outside_grid() or not self.block_fits():
                self.current_block.move(-1, 0)
                self.lock_block()

    # -------------------------------------------------------------------------

    def move_block_left(self) -> None:
        """Moves the block left 1 square."""

        if not self.game_over:
            self.current_block.move(0, -1)

            if self.is_block_outside_grid() or not self.block_fits():
                self.current_block.move(0, 1)

    # -------------------------------------------------------------------------

    def move_block_right(self) -> None:
        """Move the block right 1 square."""

        if not self.game_over:
            self.current_block.move(0, 1)

            if self.is_block_outside_grid() or not self.block_fits():
                self.current_block.move(0, -1)

    # -------------------------------------------------------------------------

    def rotate_block(self) -> None:
        """Moves the block state to its next rotation."""

        if not self.game_over:
            self.current_block.rotate_block()

            if self.is_block_outside_grid() or not self.block_fits():
                self.current_block.undo_rotation()
            else:
                self.sound.play_rotate_sound()

    # -------------------------------------------------------------------------

    def setup_window(self) -> None:
        """Window initialisation. Needs to run before anything else."""

        pr.init_window(self.window_width, self.window_height, self.title)
        pr.set_target_fps(self.targetFPS)

    # -------------------------------------------------------------------------

    def reset(self) -> None:
        """Resets the game state to the starting state."""

        self.grid.grid_setup()
        self.fill_block_list()
        self.current_block = self.get_block()
        self.next_block = self.get_block()
        self.score = 0

    # -------------------------------------------------------------------------

    def run(self) -> None:
        """All drawing and game state updates must be called here."""

        while not pr.window_should_close():
            pr.update_music_stream(self.sound.music)
            self.handle_input()
            self.automatic_move_down()

            pr.begin_drawing()

            pr.clear_background(self.background_color)
            self.draw()

            pr.end_drawing()

        self.sound.close_audio_device()
        pr.close_window()

    # -------------------------------------------------------------------------

    def snap_block_down(self) -> None:
        """Snaps the current block down as far as possible"""

        if not self.game_over:
            while True:
                self.update_score(0, 2)
                self.current_block.move(1, 0)

                if self.is_block_outside_grid() or not self.block_fits():
                    self.current_block.move(-1, 0)
                    self.lock_block()
                    break

    # -------------------------------------------------------------------------

    def update_score(self, lines_cleared: int, move_down_points: int) -> None:

        match lines_cleared:
            case 1:
                self.score += 100
            case 2:
                self.score += 300
            case 3:
                self.score += 500
            case _:
                pass

        self.score += move_down_points

    # -------------------------------------------------------------------------
